File: python/paddle/fluid/contrib/slim/core/pass_builder.py
#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserve.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from .compress_pass import CompressPass
from .config import ConfigFactory

__all__ = ['build_compressor']


def build_compressor(place=None,
                     data_reader=None,
                     data_feeder=None,
                     scope=None,
                     metrics=None,
                     epoch=None,
                     config=None):
    if config is not None:
        factory = ConfigFactory(config)
        comp_pass = factory.get_compress_pass()
    else:
        comp_pass = CompressPass()
    comp_pass.place = place
    comp_pass.data_reader = data_reader
    comp_pass.data_feeder = data_feeder
    comp_pass.scope = scope
    comp_pass.metrics = metrics
    comp_pass.epoch = epoch
    return comp_pass

File: paddle/fluid/framework/ir/pass_builder.cc
/* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License. */

#include "paddle/fluid/framework/ir/pass_builder.h"

#include <memory>

#include "glog/logging.h"
#include "paddle/fluid/platform/enforce.h"

namespace paddle::framework::ir {

class Pass;

std::shared_ptr<Pass> PassBuilder::AppendPass(const std::string& pass_type) {
  VLOG(1) << "Append " << pass_type;
  auto pass = ir::PassRegistry::Instance().Get(pass_type);
  passes_.emplace_back(pass.release());
  return passes_.back();
}

void PassBuilder::RemovePass(size_t idx) {
  PADDLE_ENFORCE_GT(
      passes_.size(),
      idx,
      common::errors::InvalidArgument(
          "Passes size is %d, %d is not a valid index.", passes_.size(), idx));
  passes_.erase(passes_.begin() + idx);  // NOLINT
}

std::shared_ptr<Pass> PassBuilder::InsertPass(size_t idx,
                                              const std::string& pass_type) {
  PADDLE_ENFORCE_GE(
      passes_.size(),
      idx,
      common::errors::InvalidArgument(
          "Passes size is %d, %d is not a valid index.", passes_.size(), idx));
  std::shared_ptr<Pass> pass(
      ir::PassRegistry::Instance().Get(pass_type).release());
  passes_.insert(passes_.begin() + idx, std::move(pass));  // NOLINT
  return passes_[idx];
}

}  // namespace paddle::framework::ir
